File: memcontainers/contracts/test/codec.test.ts
import {
  decodeExecRequest,
  decodeRelayEvent,
  encodeExecRequest,
  encodeRelayEvent,
  WireError as ControlWireError,
} from "../gen/ctl.gen.js";
import {
  decodeDefinition,
  decodeNodeDigest,
  encodeDefinition,
  encodeNodeDigest,
  WireError as LlbWireError,
  type BuildOp,
  type Definition,
  type NodeDigest,
} from "../gen/llb.gen.js";

const te = new TextEncoder();

function assert(condition: unknown, message: string): asserts condition {
  if (!condition) throw new Error(message);
}

function assertBytes(actual: Uint8Array | null | undefined, expected: Uint8Array, message: string): void {
  assert(actual !== undefined && actual !== null, `${message}: missing bytes`);
  assert(actual.length === expected.length, `${message}: length ${actual.length} !== ${expected.length}`);
  for (let i = 0; i < expected.length; i++) {
    assert(actual[i] === expected[i], `${message}: byte ${i} ${actual[i]} !== ${expected[i]}`);
  }
}

function assertEqualBytes(actual: Uint8Array, expected: Uint8Array, message: string): void {
  assertBytes(actual, expected, message);
}

function assertThrowsWire(
  fn: () => unknown,
  WireError: typeof ControlWireError | typeof LlbWireError,
  message: string,
): void {
  try {
    fn();
  } catch (error) {
    assert(error instanceof WireError, `${message}: expected ${WireError.name}, got ${(error as Error).name}`);
    return;
  }
  throw new Error(`${message}: expected throw`);
}

function u8(out: number[], value: number): void {
  out.push(value & 0xff);
}

function u16(out: number[], value: number): void {
  out.push(value & 0xff, (value >>> 8) & 0xff);
}

function u32(out: number[], value: number): void {
  out.push(value & 0xff, (value >>> 8) & 0xff, (value >>> 16) & 0xff, (value >>> 24) & 0xff);
}

function bytes(out: number[], value: Uint8Array): void {
  u32(out, value.length);
  out.push(...value);
}

function str(out: number[], value: string): void {
  bytes(out, te.encode(value));
}

function nonCanonicalExecRequestFrame(): Uint8Array {
  const out: number[] = [];
  u16(out, 1);
  u8(out, 1);
  str(out, "env");
  u8(out, 0);
  u32(out, 2);
  str(out, "z");
  str(out, "last");
  str(out, "a");
  str(out, "first");
  u8(out, 0);
  return Uint8Array.from(out);
}

function emptyBuildOp(kind: number): BuildOp {
  return { kind, parts: [], copy_paths: [], env: {}, mounts: [] };
}

function controlCodecs(): void {
  const unsorted = encodeExecRequest({
    cmd: "cat",
    cwd: "/tmp",
    env: { ZED: "z", ALPHA: "a" },
    stdin: new Uint8Array(),
  });
  const sorted = encodeExecRequest({
    cmd: "cat",
    cwd: "/tmp",
    env: { ALPHA: "a", ZED: "z" },
    stdin: new Uint8Array(),
  });
  assertEqualBytes(unsorted, sorted, "ExecRequest env map must encode canonically");

  const exec = decodeExecRequest(unsorted);
  assert(exec.cmd === "cat", "ExecRequest cmd changed");
  assert(exec.cwd === "/tmp", "ExecRequest cwd changed");
  assert(exec.env.ALPHA === "a" && exec.env.ZED === "z", "ExecRequest env changed");
  assertBytes(exec.stdin, new Uint8Array(), "ExecRequest empty stdin must stay present");

  assertThrowsWire(() => decodeExecRequest(Uint8Array.from([2, 0, 1])), ControlWireError, "wrong control message id");
  assertThrowsWire(() => decodeExecRequest(Uint8Array.from([1, 0, 2])), ControlWireError, "wrong control version");
  assertThrowsWire(() => decodeExecRequest(unsorted.slice(0, unsorted.length - 1)), ControlWireError, "truncated control frame");
  assertThrowsWire(
    () => decodeExecRequest(Uint8Array.from([...unsorted, 0])),
    ControlWireError,
    "trailing control bytes",
  );
  assertThrowsWire(
    () => decodeExecRequest(nonCanonicalExecRequestFrame()),
    ControlWireError,
    "non-canonical control strmap",
  );

  const relay = decodeRelayEvent(
    encodeRelayEvent({
      kind: "tool_approval",
      handle: 7,
      connection: "github",
      method: "POST",
      url: "https://api.example.test/repos",
      origin: "https://api.example.test",
    }),
  );
  assert(relay.kind === "tool_approval", "RelayEvent kind changed");
  assert(relay.handle === 7, "RelayEvent handle changed");
  assert(relay.args_digest === undefined, "RelayEvent absent args_digest must stay absent");

  const hostCall = decodeRelayEvent(
    encodeRelayEvent({ kind: "host_call", handle: 8, name: "empty", body: new Uint8Array() }),
  );
  assert(hostCall.name === "empty", "RelayEvent host_call name changed");
  assertBytes(hostCall.body, new Uint8Array(), "RelayEvent empty body must stay present");
}

function llbCodecs(): void {
  const source = { ...emptyBuildOp(0), source_ref: "base:latest" };
  const exec = {
    ...emptyBuildOp(7),
    input: 0,
    cmd: "printf $VALUE",
    cwd: "/work",
    env: { ZED: "z", ALPHA: "a" },
    stdin: new Uint8Array(),
    deterministic: true,
    tier: "read-write",
  };
  const definition: Definition = { version: 1, ops: [source, exec], root: 1 };
  const encoded = encodeDefinition(definition);
  const decoded = decodeDefinition(encoded);
  const reencoded = encodeDefinition(decoded);
  assertEqualBytes(reencoded, encoded, "Definition encode/decode must be canonical");
  assert(decoded.ops[1]?.cmd === "printf $VALUE", "Definition nested BuildOp changed");
  assertBytes(decoded.ops[1]?.stdin, new Uint8Array(), "Definition empty stdin must stay present");

  const unsortedDigest: NodeDigest = {
    op: exec,
    edges: [
      { role: "input", digest: "sha256:bbbb" },
      { role: "mount", digest: "sha256:aaaa" },
    ],
    resolved: { z: "last", a: "first" },
    layers: [{ digest: "sha256:cccc", size: 12, producer: "node-1" }],
    kernel_digest: "sha256:kernel",
  };
  const sortedDigest: NodeDigest = {
    ...unsortedDigest,
    op: { ...exec, env: { ALPHA: "a", ZED: "z" } },
    resolved: { a: "first", z: "last" },
  };
  assertEqualBytes(
    encodeNodeDigest(unsortedDigest),
    encodeNodeDigest(sortedDigest),
    "NodeDigest maps must encode canonically",
  );
  const digest = decodeNodeDigest(encodeNodeDigest(unsortedDigest));
  assert(digest.edges.length === 2, "NodeDigest edges list changed");
  assert(digest.layers[0]?.producer === "node-1", "NodeDigest layer list changed");

  assertThrowsWire(() => decodeDefinition(Uint8Array.from([8, 0, 1])), LlbWireError, "wrong LLB message id");
  assertThrowsWire(() => decodeDefinition(Uint8Array.from([3, 0, 2])), LlbWireError, "wrong LLB version");
  assertThrowsWire(() => decodeDefinition(encoded.slice(0, encoded.length - 1)), LlbWireError, "truncated LLB frame");
  assertThrowsWire(() => decodeDefinition(Uint8Array.from([...encoded, 0])), LlbWireError, "trailing LLB bytes");
}

controlCodecs();
llbCodecs();
console.log("contracts generated message codecs are canonical and fail closed");

File: memcontainers/contracts/test/codec_test.rs
use std::collections::BTreeMap;

use ctl_rust::{ExecRequest, RelayEvent, WireError as ControlWireError};
use llb_rust::{BuildOp, Definition, DigestEdge, LayerRef, NodeDigest, WireError as LlbWireError};

fn map(pairs: &[(&str, &str)]) -> BTreeMap<String, String> {
    pairs
        .iter()
        .map(|(k, v)| ((*k).to_string(), (*v).to_string()))
        .collect()
}

fn u8(out: &mut Vec<u8>, value: u8) {
    out.push(value);
}

fn u16(out: &mut Vec<u8>, value: u16) {
    out.extend_from_slice(&value.to_le_bytes());
}

fn u32(out: &mut Vec<u8>, value: u32) {
    out.extend_from_slice(&value.to_le_bytes());
}

fn bytes(out: &mut Vec<u8>, value: &[u8]) {
    u32(out, value.len() as u32);
    out.extend_from_slice(value);
}

fn str_bytes(out: &mut Vec<u8>, value: &str) {
    bytes(out, value.as_bytes());
}

fn non_canonical_exec_request_frame() -> Vec<u8> {
    let mut out = Vec::new();
    u16(&mut out, 1);
    u8(&mut out, 1);
    str_bytes(&mut out, "env");
    u8(&mut out, 0);
    u32(&mut out, 2);
    str_bytes(&mut out, "z");
    str_bytes(&mut out, "last");
    str_bytes(&mut out, "a");
    str_bytes(&mut out, "first");
    u8(&mut out, 0);
    out
}

#[test]
fn control_codecs_are_canonical_and_fail_closed() {
    let unsorted = ExecRequest {
        cmd: "cat".to_string(),
        cwd: Some("/tmp".to_string()),
        env: map(&[("ZED", "z"), ("ALPHA", "a")]),
        stdin: Some(Vec::new()),
    }
    .encode();
    let sorted = ExecRequest {
        cmd: "cat".to_string(),
        cwd: Some("/tmp".to_string()),
        env: map(&[("ALPHA", "a"), ("ZED", "z")]),
        stdin: Some(Vec::new()),
    }
    .encode();
    assert_eq!(unsorted, sorted);

    let exec = ExecRequest::decode(&unsorted).unwrap();
    assert_eq!(exec.cmd, "cat");
    assert_eq!(exec.cwd.as_deref(), Some("/tmp"));
    assert_eq!(exec.env.get("ALPHA").map(String::as_str), Some("a"));
    assert_eq!(exec.env.get("ZED").map(String::as_str), Some("z"));
    assert_eq!(exec.stdin.as_deref(), Some(&[][..]));

    assert_eq!(
        ExecRequest::decode(&[2, 0, 1]),
        Err(ControlWireError::WrongMessage)
    );
    assert_eq!(
        ExecRequest::decode(&[1, 0, 2]),
        Err(ControlWireError::UnsupportedVersion)
    );
    assert_eq!(
        ExecRequest::decode(&unsorted[..unsorted.len() - 1]),
        Err(ControlWireError::Truncated)
    );
    let mut trailing = unsorted.clone();
    trailing.push(0);
    assert_eq!(
        ExecRequest::decode(&trailing),
        Err(ControlWireError::TrailingBytes)
    );
    assert_eq!(
        ExecRequest::decode(&non_canonical_exec_request_frame()),
        Err(ControlWireError::NonCanonicalMap)
    );

    let relay = RelayEvent::decode(
        &RelayEvent {
            kind: "tool_approval".to_string(),
            handle: 7,
            connection: Some("github".to_string()),
            method: Some("POST".to_string()),
            url: Some("https://api.example.test/repos".to_string()),
            origin: Some("https://api.example.test".to_string()),
            ..Default::default()
        }
        .encode(),
    )
    .unwrap();
    assert_eq!(relay.kind, "tool_approval");
    assert_eq!(relay.handle, 7);
    assert_eq!(relay.args_digest, None);

    let host_call = RelayEvent::decode(
        &RelayEvent {
            kind: "host_call".to_string(),
            handle: 8,
            name: Some("empty".to_string()),
            body: Some(Vec::new()),
            ..Default::default()
        }
        .encode(),
    )
    .unwrap();
    assert_eq!(host_call.name.as_deref(), Some("empty"));
    assert_eq!(host_call.body.as_deref(), Some(&[][..]));
}

#[test]
fn llb_codecs_are_canonical_and_fail_closed() {
    let source = BuildOp {
        kind: 0,
        source_ref: Some("base:latest".to_string()),
        ..Default::default()
    };
    let exec = BuildOp {
        kind: 7,
        input: Some(0),
        cmd: Some("printf $VALUE".to_string()),
        cwd: Some("/work".to_string()),
        env: map(&[("ZED", "z"), ("ALPHA", "a")]),
        stdin: Some(Vec::new()),
        deterministic: Some(true),
        tier: Some("read-write".to_string()),
        ..Default::default()
    };
    let definition = Definition {
        version: 1,
        ops: vec![source, exec.clone()],
        root: 1,
    };
    let encoded = definition.encode();
    let decoded = Definition::decode(&encoded).unwrap();
    assert_eq!(decoded.encode(), encoded);
    assert_eq!(decoded.ops[1].cmd.as_deref(), Some("printf $VALUE"));
    assert_eq!(decoded.ops[1].stdin.as_deref(), Some(&[][..]));

    let unsorted_digest = NodeDigest {
        op: exec.clone(),
        edges: vec![
            DigestEdge {
                role: "input".to_string(),
                digest: "sha256:bbbb".to_string(),
            },
            DigestEdge {
                role: "mount".to_string(),
                digest: "sha256:aaaa".to_string(),
            },
        ],
        resolved: map(&[("z", "last"), ("a", "first")]),
        layers: vec![LayerRef {
            digest: "sha256:cccc".to_string(),
            size: 12,
            producer: "node-1".to_string(),
        }],
        kernel_digest: Some("sha256:kernel".to_string()),
    };
    let mut sorted_exec = exec;
    sorted_exec.env = map(&[("ALPHA", "a"), ("ZED", "z")]);
    let sorted_digest = NodeDigest {
        op: sorted_exec,
        resolved: map(&[("a", "first"), ("z", "last")]),
        ..unsorted_digest.clone()
    };
    assert_eq!(unsorted_digest.encode(), sorted_digest.encode());
    let digest = NodeDigest::decode(&unsorted_digest.encode()).unwrap();
    assert_eq!(digest.edges.len(), 2);
    assert_eq!(digest.layers[0].producer, "node-1");

    assert_eq!(
        Definition::decode(&[8, 0, 1]),
        Err(LlbWireError::WrongMessage)
    );
    assert_eq!(
        Definition::decode(&[3, 0, 2]),
        Err(LlbWireError::UnsupportedVersion)
    );
    assert_eq!(
        Definition::decode(&encoded[..encoded.len() - 1]),
        Err(LlbWireError::Truncated)
    );
    let mut trailing = encoded;
    trailing.push(0);
    assert_eq!(
        Definition::decode(&trailing),
        Err(LlbWireError::TrailingBytes)
    );
}
